magpie/models.py

```diff
@@ -132,14 +145,309 @@ def __str__(self):
+        those models should be accessed from the result.
+    """
+
     @classmethod
-    def search(cls, status=None, db_session=None):
+    def by_status(cls, status=None, db_session=None):
+        # type: (Optional[UserStatuses], Optional[Session]) -> Iterable[AnyUser]
+        """
+        Search for appropriate :class:`User` and/or :class:`UserPending` according to specified :class:`UserStatuses`.
+
+        When the :paramref:`status` is ``None``, *normal* retrieval of all non-pending :class:`User` is executed, as if
+        directly using the :class:`UserService` implementation.
+        Otherwise, a combination of appropriate search criterion is executed based on the :paramref:`status` flags.
+        """
         db_session = get_db_session(db_session)
+        if status is UserStatuses.Pending:
+            return db_session.query(UserPending)
+        if status is None:
+            # original behavior, all approved users returned regardless of statuses, ignoring pending ones
+            return db_session.query(cls.model)
         query = db_session.query(cls.model)
-        if status is not None:
-            query = query.filter(cls.model.status == status)
-        return query
+        users = []  # must combine by list since different models will clash in query
+        if UserStatuses.Pending in status:
+            users = list(db_session.query(UserPending))
+            status = UserStatuses(status - UserStatuses.Pending)
+        status = [int(status_flag) for status_flag in status]
+        query = query.filter(cls.model.status.in_(status))
+        users += list(query)
+        return users
+
+    @classmethod
+    def by_user_name(cls, user_name, status=None, db_session=None):  # pylint: disable=W0221,arguments-differ
+        # type: (Str, Optional[UserStatuses], Optional[Session]) -> Optional[AnyUser]
+        """
+        Retrieves the user matching the given name.
+
+        Search is always accomplished against :class:`User` table unless :attr:`UserStatuses.Pending` is provided in
+        the :paramref:`status`. If more that one status is provided such that both :class:`UserPending` and
+        :class:`User` could yield results, the :class:`User` is returned first, as there should not be any conflict
+        between those two models.
+        """
+        if status is None or UserStatuses.Pending not in status:
+            return super(UserSearchService, cls).by_user_name(user_name, db_session=db_session)
+        users = list(cls.by_status(status=status, db_session=db_session))
+        if not users:
+            return None
+        users = [user for user in users if user.user_name == user_name]
+        if not users:
+            return None
+        if len(users) == 1:
+            return users[0]
+        LOGGER.warning("Duplicate registered/pending users named [%s] detected! This should never happen.", user_name)
+        return users[0] if isinstance(users[0], User) else users[1]
```

Comment:
Having more than one user with the same name is indeed suspicious, but having something else at [0]?!? What could be there apart from a `User`?

Reply:
This class handles the 2 kind of `User` / `UserPending` entities. 
Technically, using a combination of statutes (`UserStatuses.OK|UserStatuses.Pending` or `"all"`) will run a combined search on the 2 tables, which *could* return distinct entities, with duplicate names. 

I say *technically*, because one would have to manually inject a duplicate name in the other user-type table since each table individually blocks duplicate names with unique constraint. The API also blocks the creation when a duplicate is found in any table. The `[0]` vs `[1]` is in that case is to prioritize the `User` definition over the incomplete `UserPending` one, if that ever happened. Only possible results are `[None, None]`, `[None, User]`, `[None, UserPending]`, `[User, UserPending]` (this edge case) and their swapped around versions since unique are guaranteed from both tables. 

The only way I can think of to guarantee non-duplicate, would be to have a specific `user_name` table with unique constraint, with both `User` and `UserPending` pointing to it by foreign key for their respective names. The only issue about this is that `User` is implemented by ziggurat, and I will need to fiddle with their model. Could work, but could break things unexpectedly somewhere.

Reply:
Would you mind adding that explanation above the return statement? It makes perfect sense and I don't see any requirements to add a unique username table.

Reply:
In fact the function main documentation explains that well. I was only trigerred by the warning message just above. Forgot my previous comment.